collator/src/collator/do_collate/finalize.rs

```diff
@@ -555,6 +593,34 @@ impl Phase<FinalizeState> {
                     self.state.collation_data.processed_upto.externals.as_ref(),
                 );
 
+                let mut shards_processed_to = FastHashMap::default();
+
+                for (shard_id, shard_data) in shards.iter() {
+                    if !shard_data.top_sc_block_updated {
```

Comment:
Looks like we will totally skip shards that were not updated instead of reading info from the previous `McData`.

Reply:
I suggest to write a helper function and make a unit test.